Server-only authentication of backends and per-location SSL config

janosi opened the pull request:
**What this PR does / why we need it**:
1. It enables the one-way authentication of backends based on the ca.crt configured in the annotation proxy-ssl-secret. 
2. It enables the configuration of different proxy-ssl-secrets for the different locations of the same server

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #4688 #4503 

**Special notes for your reviewer**:

aledbf commented:
/ok-to-test

aledbf commented:
/retest

janosi commented:
/assign @bowei

aledbf commented on internal/ingress/controller/controller.go:
```diff
@@ -491,17 +491,6 @@ func (n *NGINXController) getBackendServers(ingresses []*ingress.Ingress) ([]*in
 					server.Hostname, ingKey)
 			}
 
-			if server.ProxySSL.CAFileName == "" {
```
Why are you removing this?
Without it, you can overwrite the server certificate, if the annotation is used in two different Ingress rules for the same hostname.

janosi replied:
Exactly that is the purpose here. The intent is to move the proxy-ssl rules to location level instead of the current server level. So users can then define certificates on location (backend/service) level. I.e. the different locations under the same server can have different certificates. This is an ingress usage pattern in our cloud.

aledbf commented on internal/ingress/controller/store/backend_ssl.go:
```diff
@@ -104,17 +104,19 @@ func (s *k8sStore) getPemCertificate(secretName string) (*ingress.SSLCert, error
 			return nil, fmt.Errorf("unexpected error creating SSL Cert: %v", err)
 		}
 
+		path, err := ssl.StoreSSLCertOnDisk(nsSecName, sslCert)
```
Why are you doing this?
Only CA certificates should be stored on disk. Everything else is handled by Lua, dynamically.

janosi replied:
When I used the original version and I defined proxy-ssl-secret in my Ingress the result was a config file in which all of these: proxy_ssl_trusted_cert, proxy_ssl_certificate, proxy_ssl_certificate_key pointed to the same file on disk. It was my experience with the original version. I.e. it wanted to have the tls.crt and tls.key from a file on a disk originally.
If I understand the original code well it does the following: puts the client cert and key into sslCert, then if server CA is present then it first stores the sslCert (i.e. client key and cert) on disk first, then it appends the server CA to the file.
In order to preserve this function I kept the writing the client key and cert to disk in my code.  I tested it, it works. Both when there is only ca.crt in the proxy-ssl-secret, and also when there are additional tls.key and tls.crt.

janosi replied:
@aledbf I would like to ask, could you please help me in finding that lua script that is supposed to manage the client key and cert? I would like to see the whole picture (even though my experience with the original version showed that the controller configured nginx to take those from the CAFile from disk) Thank you!

janosi commented:
@aledbf Is there anything I can do for this PR?

aledbf commented:
>  Is there anything I can do for this PR?

Don't break the current behavior. I understand your goal here but this change will break existing ingresses after an upgrade.

janosi commented:
@aledbf 
I do not think it breaks any behavior. The proxy-ssl parameters are moved to the location level from server level, but those are still there and applied on every location under the same server. I.e .the functionality is the same.
Also about the storing the secrets on disk: it is not new here, the only thing I did is that I moved the disk writing out from the condition that checks whether the ca.crt is available or  not (and I think ca.crt is always available, as that is a must for the minimal server/backend authentication - so the tls.key and tls.crt were always written to disk in reality). So, writing the tls.key and tls.crt to disk is not a new thing that I introduced here.
Please help me in understanding what it breaks from functionality perspective.
On the other hand it fixes the problem that people currently cannot use one-way auth to the backends, and that people currently cannot have different secrets for the different paths.

janosi commented:
@aledbf I restored the functionality that puts the proxy_ssl parameters on server level. I.e. the old functionality is restored. Beside that the new logic puts the proxy_ssl parameters to location level, _too_.
Luckily if both a server and a corresponding location have proxy_ssl parameters the parameters of the location are taken into account - i.e. the missing/incorrect functionality is fixed now, while preserving the old logic. 
Please check.

janosi commented:
@aledbf Could you please check and comment? Thank you!

janosi commented:
@aledbf @ElvinEfendi Could any of you please check this? Thank you!

aledbf commented:
/lgtm

janosi commented:
@aledbf Thank you!

EronWright commented:
For those who find themselves here, this PR does make it possible to use a secret containing only a CA certificate bundle, as referenced by the `ginx.ingress.kubernetes.io/proxy-ssl-secret` annotation.  Thanks again @janosi!

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: proxy-ssl-secret
data:
  ca.crt: (base64-encoded PEM file)
type: Opaque
```
